### DISCUSSION TITLE
What would be simplest way to add CLI help reference into sphinx documentation?

### DISCUSSION THREAD

**Discussion by Tobi-De** [DanCardin/cappa#87] [Q&A] [answered]
Hello there, I hope you are doing great.

If I need to include the help message of my commands automatically inside a sphinx documentation, or generate documentation from my cli app, what are my options ?

**Comment by DanCardin** [comment]
pawamoy had a similar usecase, but for mkdocs: https://github.com/DanCardin/cappa/discussions/69. Wherein he produced markdown from the cli structure and presumably exec'd the result somewhere else, although i can't say the details.

Since it's come up twice in a relatively small sample set of people, i'd be interested in a first-class feature for supporting the usecase, but it's not obvious to me at this point what form that would take. Perhaps at least for sphinx that'd mean a sphinx plugin.

**Reply by Tobi-De** [reply]
There is a lot of good options available for mkdocs and markdown (ex: https://github.com/ewels/rich-codex), but I can't find what I want with sphinx and rst, there is https://github.com/WhyNotHugo/sphinx-autorun but it ia barely maintained and do not produce colored output 😞

**Reply by pawamoy** [reply]
You can probably rework the snippet in #69 as a custom Sphinx directive. Add a `cappa_cli.py` module in your `docs` folder, make sure it's exposed in Python's `sys.path` (by adding the `docs` folder to it in `conf.py` for example), and add `"cappa_cli"` to the `extensions` list in `conf.py`.

```tree
docs/
    conf.py
    cappa_cli.py
```

```python
extensions = [
    "cappa_cli",
]
```

In `cappa_cli.py`, register your directives like so:

```python
from docutils.parsers.rst import Directive


class CappaCli(Directive):
    ...


def setup(app):
    app.add_directive("cappa_cli", CappaCli)
```

[The docs](https://www.sphinx-doc.org/en/master/development/overview.html) will explain what to put in `CappaCli` better than I would.

Then you can inject your CLI docs in a rST document with this:

```rst
.. cappa_cli::
    :cli_class: package.module:CappaDataclass

.. the option above just shows what the directive could accept
```

(You might already know all this, just writing for reference)

---

Looking at the above, it's probably not the *simplest* way. Maybe you could also use the [programoutput](https://pythonhosted.org/sphinxcontrib-programoutput/) extension, though I think you'd have to output HTML directly, which is not great.

**Reply by DanCardin** [reply, marked as answer]
I've started looking into a sphinx directive I could ship with the library. seems worth, even to my own docs.

**Reply by DanCardin** [reply]
@Tobi-De https://github.com/DanCardin/cappa/pull/89. If you have a chance, ideally you could try this out before I merge it.

**Reply by Tobi-De** [reply]
> @Tobi-De https://github.com/DanCardin/cappa/pull/89. If you have a chance, ideally you could try this out before I merge it.

Nice, I will

**Reply by Tobi-De** [reply]
I might be doing something wrong, but right now it shows nothing

```rst
Start project
=============

.. cappa falco.commands.StartProject

```
the sphinx build command shows not errors, the extensions is loaded but the directive does nothing
I'm testing it agains this repo https://github.com/Tobi-De/falco/blob/main/docs/the_cli/start_project.rst
https://github.com/Tobi-De/falco/blob/main/src/falco/commands/start_project.py (the command I tried)

**Reply by Tobi-De** [reply]
my bad, there is something missing in my directive, I copied it from yours docs, so it is missing there too, the correct syntax is

```rst
.. cappa:: falco.commands.StartProject
```

**Reply by DanCardin** [reply]
woops! will update the docs. i use MyST nowadays, so i forget the rst syntaxes

**Reply by Tobi-De** [reply]
<img width="1571" alt="Screen Shot 2023-12-18 at 17 14 58" src="https://github.com/DanCardin/cappa/assets/40334729/bd55ff09-21d3-4ec1-a18d-61f8c5d1a127">

The style is broken, same thing when I set the style to `terminal`

**Reply by Tobi-De** [reply]
the terminal style make no difference btw

**Reply by DanCardin** [reply]
I forgot to upload my example screenshots to the PR (updated), but adding here also:

<img width="792" alt="Screenshot 2023-12-18 at 9 39 40 AM" src="https://github.com/DanCardin/cappa/assets/701548/9a9e7dd6-836e-4c33-ab8c-33c71502b50f">

What I think might be happening actually, is that it's doing hard line wraps. My guess is your terminal is very wide and leading to it not wrapping early enough. Not sure what the best solution is here, since soft line wraps wont wrap in the same way they would in the terminal. Likely the solution will need to be a setting to control the "terminal" width.

---

Not sure what you mean by

> the terminal style make no difference btw

**Reply by DanCardin** [reply]
I'm not seeing the behavior you're getting no matter the actual width though. I wonder if it's a specific behavior of the CSS of your theme.

As far as addressing the issue generally, pushed an update that adds allows setting `:terminal-width: 80`.

However it defaults to "0", which does browser-level line wrapping. The caveat is that it might wrap in ways that are not quite consistent with how it'd wrap in the terminal. But it seemed less opinionated than defaulting to a specific width, and perhaps that not even fixing your visual issue, depending on the browser width.

**Reply by Tobi-De** [reply]
> Not sure what you mean by

There is no visual difference between

```rst
.. cappa:: falco.commands.StartProject
```

and this

```rst
.. cappa:: falco.commands.StartProject
	:style: terminal
```

**Reply by DanCardin** [reply]
Right. perhaps that's not clear in my docs prose. the default is `terminal`. So those are exactly equivalent

**Reply by Tobi-De** [reply]
what are the available options?

**Reply by Tobi-De** [reply]
I tried with the new change and it's better, but it's a bit too shrunken, I tried adjusting the terminal width and it worked but it didn't resize on small screens, I assume it's some kind of fixed width, would it be possible to make it some kind of max-width?

**Reply by DanCardin** [reply]
the default, if left unspecified should be `:terminal-width: 0`, which leaves the wrapping to the browser (but doesnt line-wrap identically to the actual terminal).

The fixed width option, if you set it, is fixed width because it's doing literal hard line-wrapping. But it **should** be scrolling horizontally inside the `pre`/`code` block. At least that's the behavior I see with my own documentation/theme.

**Reply by Tobi-De** [reply]
I see, I think you can release it as is, it is already a great starting point, I would probably have to do some tweaking related to my theme, thanks for your work 😄,
